ios/PurchasesHybridCommon/PurchasesHybridCommonSwift/CommonFunctionality.swift

```diff
@@ -0,0 +1,505 @@
+        Purchases.logLevel = enabled ? .debug : .info
+    }
+
+    @objc public static func setAutomaticAppleSearchAdsAttributionCollection(_ enabled: Bool) {
+        Purchases.automaticAppleSearchAdsAttributionCollection = enabled
+    }
+
+    @objc public static func setFinishTransactions(_ finishTransactions: Bool) {
+        Purchases.shared.finishTransactions = finishTransactions
+    }
+
+    @objc public static func invalidatePurchaserInfoCache() {
+        Purchases.shared.invalidatePurchaserInfoCache()
+    }
+
+    @available(iOS 14.0, *)
+    @available(tvOS, unavailable)
+    @available(macOS, unavailable)
+    @available(watchOS, unavailable)
+    @objc public static func presentCodeRedemptionSheet() {
+        Purchases.shared.presentCodeRedemptionSheet()
+    }
+
+    @objc public static func canMakePaymentsWithFeatures(_ features: [Int]) -> Bool {
+        return Purchases.canMakePayments()
+    }
+
+}
+
+// MARK: purchasing and restoring
+@objc public extension CommonFunctionality {
+
+    @objc(restoreTransactionsWithCompletionBlock:)
+    static func restoreTransactions(completion: @escaping ([String: Any]?, ErrorContainer?) -> Void) {
+        let purchaserInfoCompletion = purchaserInfoCompletionBlock(from: completion)
+        Purchases.shared.restoreTransactions(purchaserInfoCompletion)
+    }
+
+    @objc(syncPurchasesWithCompletionBlock:)
+    static func syncPurchases(completion: (([String: Any]?, ErrorContainer?) -> Void)?) {
+        if let completion = completion {
+            let purchaserInfoCompletion = purchaserInfoCompletionBlock(from: completion)
+            Purchases.shared.restoreTransactions(purchaserInfoCompletion)
+        } else {
+            Purchases.shared.restoreTransactions(nil)
+        }
+    }
+
+    @objc(purchaseProduct:signedDiscountTimestamp:completionBlock:)
+    static func purchaseProduct(_ productIdentifier: String,
+                                signedDiscountTimestamp: String?,
+                                completion: @escaping ([String: Any]?, ErrorContainer?) -> Void) {
+        let hybridCompletion: (SKPaymentTransaction?,
+                               Purchases.PurchaserInfo?,
+                               Error?,
+                               Bool) -> Void = { transaction, purchaserInfo, error, userCancelled in
+            if let error = error {
+                completion(nil, ErrorContainer(error: error, extraPayload: ["userCancelled": userCancelled]))
+            } else if let purchaserInfo = purchaserInfo,
+                      let transaction = transaction {
+                completion([
+                    "purchaserInfo": purchaserInfo.dictionary,
+                    "productIdentifier": transaction.payment.productIdentifier
+                ], nil)
+            } else {
+                // todo: maybe set up custom error here?
+                completion(nil, productNotFoundError(description: "Couldn't find product.", userCancelled: false))
```

Comment:
swift nullability checks uncovered a few edge cases here and there that we weren't covering before